pkg/export/secrets/manager.go

```diff
@@ -0,0 +1,223 @@
+type SecretConfig struct {
+	Name   string                 `yaml:"name"`
+	Config KubernetesSecretConfig `yaml:"config"`
+}
+
+type secretEntry struct {
+	config KubernetesSecretConfig
+	secret Secret
+}
+
+// Manager manages the Kubernetes secret provider.
+type Manager struct {
+	ctx      context.Context
+	cancelFn func()
+	provider *watchProvider
+	config   *WatchSPConfig
+	secrets  map[string]*secretEntry
+	opts     ProviderOptions
+}
+
+// NewManager creates a new secret manager with the provided options.
+func NewManager(ctx context.Context, reg prometheus.Registerer, opts ProviderOptions) Manager {
+	if reg != nil {
+		reg.MustRegister(failedSecretConfigs)
+		reg.MustRegister(secretsTotal)
+	}
+	return Manager{
+		ctx:      ctx,
+		cancelFn: func() {},
+		secrets:  make(map[string]*secretEntry),
+		opts:     opts,
+	}
+}
+
+// ApplyConfig applies the new secrets, diffing each one with the last configuration to apply the
+// relevant update.
+func (m *Manager) ApplyConfig(ctx context.Context, providerConfig *WatchSPConfig, configs []SecretConfig) error {
+	// If no secrets are provided, cancel any existing secret provider.
+	if len(configs) == 0 {
+		m.cancelFn()
+		m.provider = nil
+		m.cancelFn = func() {}
+		m.secrets = map[string]*secretEntry{}
+		m.config = nil
+		return nil
+	}
+
+	eq, err := yamlEqual(m.config, providerConfig)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		m.config = providerConfig
+	}()
+
+	if !eq || m.provider == nil {
+		ctx, cancel := context.WithCancel(m.ctx)
+		provider, err := providerConfig.newProvider(ctx, m.opts)
+		if err != nil {
+			cancel()
+			return err
+		}
+
+		m.cancelFn()
+		m.provider = provider
+		m.cancelFn = cancel
+		m.secrets = map[string]*secretEntry{}
+	}
+	return m.updateSecrets(ctx, configs)
+}
+
+func (m *Manager) updateSecrets(ctx context.Context, configs []SecretConfig) error {
+	var errs []error
+	secretNamesEnabled := make(map[string]bool)
+	for _, secret := range configs {
+		if enabled, ok := secretNamesEnabled[secret.Name]; ok {
+			if !enabled {
+				continue
+			}
+			errs = append(errs, fmt.Errorf("duplicate secret key %q", secret.Name))
+			secretNamesEnabled[secret.Name] = false
+		} else {
+			secretNamesEnabled[secret.Name] = true
+		}
+	}
+
+	newSecrets := map[string]*secretEntry{}
+	for i := range configs {
+		secret := &configs[i]
+		if enabled := secretNamesEnabled[secret.Name]; !enabled {
+			continue
+		}
+		if secretConfig, ok := m.secrets[secret.Name]; ok {
+			eq, err := yamlEqual(&secretConfig.config, &secret.Config)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			delete(m.secrets, secret.Name)
+			if eq {
+				newSecrets[secret.Name] = secretConfig
+				continue
+			}
+			s, err := m.provider.Update(ctx, &secretConfig.config, &secret.Config)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			secretConfig.secret = s
+			newSecrets[secret.Name] = secretConfig
+			continue
```

Comment:
nit: use else instead of continue. Reads better. 
if it exists{ 
// modify
} else {
// add it
}